File: apps/iris/src-tauri/plugins/mobile-browser/android/src/main/java/to/iris/browser/mobile/MobileBrowserPlugin.kt
package to.iris.browser.mobile

import android.app.Activity
import android.graphics.Bitmap
import android.graphics.Color
import android.net.http.SslError
import android.net.Uri
import android.util.Log
import android.util.TypedValue
import android.view.ViewGroup
import android.webkit.WebChromeClient
import android.webkit.WebResourceError
import android.webkit.WebResourceRequest
import android.webkit.SslErrorHandler
import android.webkit.WebSettings
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.FrameLayout
import androidx.core.content.ContextCompat
import androidx.webkit.WebViewCompat
import androidx.webkit.WebViewFeature
import app.tauri.annotation.Command
import app.tauri.annotation.InvokeArg
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.Invoke
import app.tauri.plugin.JSObject
import app.tauri.plugin.Plugin
import kotlin.math.roundToInt

@InvokeArg
class CreateArgs {
    lateinit var label: String
    lateinit var url: String
    var x: Double = 0.0
    var y: Double = 0.0
    var width: Double = 0.0
    var height: Double = 0.0
    var scale: Double = 1.0
    lateinit var initScript: String
    lateinit var diagnosticScript: String
    var allowedOriginRule: String? = null
    var actualUrlRoot: String? = null
}

@InvokeArg
class LabelArgs {
    lateinit var label: String
}

@InvokeArg
class NavigateArgs {
    lateinit var label: String
    lateinit var url: String
}

@InvokeArg
class BoundsArgs {
    lateinit var label: String
    var x: Double = 0.0
    var y: Double = 0.0
    var width: Double = 0.0
    var height: Double = 0.0
    var scale: Double = 1.0
}

@InvokeArg
class ShellOverlayArgs {
    var enabled: Boolean = false
    var x: Double = 0.0
    var y: Double = 0.0
    var width: Double = 0.0
    var height: Double = 0.0
    var scale: Double = 1.0
}

@InvokeArg
class HistoryArgs {
    lateinit var label: String
    lateinit var direction: String
}

private data class BrowserEntry(
    val args: CreateArgs,
    val webView: WebView,
)

@TauriPlugin
class MobileBrowserPlugin(private val activity: Activity) : Plugin(activity) {
    companion object {
        private const val TAG = "MobileBrowserPlugin"
    }

    private val browsers = mutableMapOf<String, BrowserEntry>()

    @Command
    fun create(invoke: Invoke) {
        val args = invoke.parseArgs(CreateArgs::class.java)
        activity.runOnUiThread {
            destroyBrowser(args.label)

            val webView = WebView(activity)
            configureWebView(webView, args)
            browsers[args.label] = BrowserEntry(args, webView)

            val root = contentRoot()
            root.setBackgroundColor(resolveThemeBackgroundColor())
            root.addView(webView, layoutParams(args))
            root.bringChildToFront(webView)
            webView.bringToFront()
            root.requestLayout()
            root.invalidate()
            Log.i(TAG, "create ${args.label} url=${args.url} bounds=${describe(args.x, args.y, args.width, args.height, args.scale)}")
            webView.loadUrl(args.url)
            invoke.resolve()
        }
    }

    @Command
    fun close(invoke: Invoke) {
        val args = invoke.parseArgs(LabelArgs::class.java)
        activity.runOnUiThread {
            destroyBrowser(args.label)
            invoke.resolve()
        }
    }

    @Command
    fun navigate(invoke: Invoke) {
        val args = invoke.parseArgs(NavigateArgs::class.java)
        activity.runOnUiThread {
            val browser = browsers[args.label]?.webView
            if (browser == null) {
                invoke.reject("Webview ${args.label} not found")
                return@runOnUiThread
            }
            Log.i(TAG, "navigate ${args.label} url=${args.url}")
            browser.loadUrl(args.url)
            invoke.resolve()
        }
    }

    @Command
    fun setBounds(invoke: Invoke) {
        val args = invoke.parseArgs(BoundsArgs::class.java)
        activity.runOnUiThread {
            val browser = browsers[args.label]?.webView
            if (browser == null) {
                invoke.reject("Webview ${args.label} not found")
                return@runOnUiThread
            }
            browser.layoutParams = layoutParams(args)
            contentRoot().bringChildToFront(browser)
            browser.bringToFront()
            browser.requestLayout()
            Log.i(TAG, "setBounds ${args.label} bounds=${describe(args.x, args.y, args.width, args.height, args.scale)}")
            invoke.resolve()
        }
    }

    @Command
    fun setShellOverlay(invoke: Invoke) {
        invoke.parseArgs(ShellOverlayArgs::class.java)
        activity.runOnUiThread {
            invoke.resolve()
        }
    }

    @Command
    fun history(invoke: Invoke) {
        val args = invoke.parseArgs(HistoryArgs::class.java)
        activity.runOnUiThread {
            val browser = browsers[args.label]?.webView
            if (browser == null) {
                invoke.reject("Webview ${args.label} not found")
                return@runOnUiThread
            }
            when (args.direction) {
                "back" -> if (browser.canGoBack()) browser.goBack()
                "forward" -> if (browser.canGoForward()) browser.goForward()
                else -> {
                    invoke.reject("Invalid history direction")
                    return@runOnUiThread
                }
            }
            invoke.resolve()
        }
    }

    @Command
    fun reload(invoke: Invoke) {
        val args = invoke.parseArgs(LabelArgs::class.java)
        activity.runOnUiThread {
            val browser = browsers[args.label]?.webView
            if (browser == null) {
                invoke.reject("Webview ${args.label} not found")
                return@runOnUiThread
            }
            browser.reload()
            invoke.resolve()
        }
    }

    @Command
    fun currentUrl(invoke: Invoke) {
        val args = invoke.parseArgs(LabelArgs::class.java)
        activity.runOnUiThread {
            val browser = browsers[args.label]?.webView
            if (browser == null) {
                invoke.reject("Webview ${args.label} not found")
                return@runOnUiThread
            }
            val payload = JSObject()
            payload.put("url", browser.url)
            invoke.resolve(payload)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        activity.runOnUiThread {
            browsers.keys.toList().forEach(::destroyBrowser)
        }
    }

    private fun configureWebView(webView: WebView, args: CreateArgs) {
        with(webView.settings) {
            javaScriptEnabled = true
            domStorageEnabled = true
            mediaPlaybackRequiresUserGesture = false
            javaScriptCanOpenWindowsAutomatically = true
            mixedContentMode = WebSettings.MIXED_CONTENT_ALWAYS_ALLOW
        }

        webView.setBackgroundColor(resolveThemeBackgroundColor())
        webView.webChromeClient = WebChromeClient()

        if (WebViewFeature.isFeatureSupported(WebViewFeature.DOCUMENT_START_SCRIPT)) {
            val allowedOrigins = setOf(args.allowedOriginRule ?: "*")
            WebViewCompat.addDocumentStartJavaScript(webView, args.initScript, allowedOrigins)
        }

        webView.webViewClient = object : WebViewClient() {
            override fun shouldOverrideUrlLoading(view: WebView, request: WebResourceRequest): Boolean {
                if (!request.isForMainFrame) return false
                val url = request.url.toString()
                return shouldInterceptNavigation(args, url)
            }

            override fun onPageStarted(view: WebView, url: String?, favicon: Bitmap?) {
                super.onPageStarted(view, url, favicon)
                val safeUrl = url ?: return
                Log.i(TAG, "page started ${args.label} url=$safeUrl")
                emitLocation(args.label, safeUrl, "navigation")
                emitPageLoad(args.label, safeUrl, "started")
            }

            override fun onPageFinished(view: WebView, url: String?) {
                super.onPageFinished(view, url)
                val safeUrl = url ?: return
                Log.i(TAG, "page finished ${args.label} url=$safeUrl")
                if (shouldInjectScripts(args, safeUrl)) {
                    injectScripts(view, args)
                }
                emitPageLoad(args.label, safeUrl, "finished")
            }

            override fun onReceivedError(
                view: WebView,
                request: WebResourceRequest,
                error: WebResourceError,
            ) {
                super.onReceivedError(view, request, error)
                if (!request.isForMainFrame) return
                val failingUrl = request.url.toString()
                val description = error.description?.toString() ?: "Failed to load page"
                Log.e(TAG, "page error ${args.label} url=$failingUrl code=${error.errorCode} error=$description")
                emitDiagnostic(args.label, failingUrl, "page-load-error", description)
            }

            override fun onReceivedSslError(
                view: WebView,
                handler: SslErrorHandler,
                error: SslError,
            ) {
                super.onReceivedSslError(view, handler, error)
                val failingUrl = error.url ?: view.url ?: args.url
                val description = "TLS error ${error.primaryError}"
                Log.e(TAG, "ssl error ${args.label} url=$failingUrl error=$description")
                emitDiagnostic(args.label, failingUrl, "ssl-error", description)
            }
        }
    }

    private fun shouldInterceptNavigation(args: CreateArgs, url: String): Boolean {
        if (url.startsWith("htree://")) {
            emitLocation(args.label, url, "navigation")
            return true
        }

        if (args.actualUrlRoot != null) {
            if (!url.startsWith(args.actualUrlRoot!!)) {
                emitLocation(args.label, url, "navigation")
                return true
            }
            return false
        }

        val allowedOrigin = args.allowedOriginRule ?: return false
        if (originOf(url) != allowedOrigin) {
            emitLocation(args.label, url, "navigation")
            return true
        }

        return false
    }

    private fun shouldInjectScripts(args: CreateArgs, url: String): Boolean {
        if (args.actualUrlRoot != null) {
            return url.startsWith(args.actualUrlRoot!!)
        }

        val allowedOrigin = args.allowedOriginRule ?: return true
        return originOf(url) == allowedOrigin
    }

    private fun injectScripts(webView: WebView, args: CreateArgs) {
        webView.evaluateJavascript(args.initScript, null)
        webView.evaluateJavascript(args.diagnosticScript, null)
        webView.postDelayed({ webView.evaluateJavascript(args.initScript, null) }, 150)
        webView.postDelayed({ webView.evaluateJavascript(args.initScript, null) }, 1000)
        webView.postDelayed({ webView.evaluateJavascript(args.diagnosticScript, null) }, 150)
        webView.postDelayed({ webView.evaluateJavascript(args.diagnosticScript, null) }, 1000)
    }

    private fun emitLocation(label: String, url: String, source: String) {
        val payload = JSObject()
        payload.put("label", label)
        payload.put("url", url)
        payload.put("source", source)
        trigger("location", payload)
    }

    private fun emitPageLoad(label: String, url: String, event: String) {
        val payload = JSObject()
        payload.put("label", label)
        payload.put("url", url)
        payload.put("event", event)
        trigger("page-load", payload)
    }

    private fun emitDiagnostic(label: String, url: String?, source: String, error: String) {
        val payload = JSObject()
        payload.put("label", label)
        payload.put("url", url)
        payload.put("source", source)
        payload.put("error", error)
        trigger("diagnostic", payload)
    }

    private fun destroyBrowser(label: String) {
        val entry = browsers.remove(label) ?: return
        (entry.webView.parent as? ViewGroup)?.removeView(entry.webView)
        entry.webView.stopLoading()
        entry.webView.webChromeClient = null
        entry.webView.webViewClient = WebViewClient()
        entry.webView.removeAllViews()
        entry.webView.destroy()
    }

    private fun contentRoot(): FrameLayout {
        return activity.findViewById(android.R.id.content)
    }

    private fun layoutParams(args: CreateArgs): FrameLayout.LayoutParams {
        return FrameLayout.LayoutParams(
            nativePx(args.width, args.scale),
            nativePx(args.height, args.scale),
        ).apply {
            leftMargin = nativePx(args.x, args.scale)
            topMargin = nativePx(args.y, args.scale)
        }
    }

    private fun layoutParams(args: BoundsArgs): FrameLayout.LayoutParams {
        return FrameLayout.LayoutParams(
            nativePx(args.width, args.scale),
            nativePx(args.height, args.scale),
        ).apply {
            leftMargin = nativePx(args.x, args.scale)
            topMargin = nativePx(args.y, args.scale)
        }
    }

    private fun nativePx(value: Double, scale: Double): Int {
        val safeScale = if (scale.isFinite() && scale > 0.0) scale else 1.0
        return (value * safeScale).roundToInt().coerceAtLeast(0)
    }

    private fun describe(x: Double, y: Double, width: Double, height: Double, scale: Double): String {
        return "css=($x,$y ${width}x$height) scale=$scale native=(${nativePx(x, scale)},${nativePx(y, scale)} ${nativePx(width, scale)}x${nativePx(height, scale)})"
    }

    private fun originOf(url: String): String? {
        return try {
            val uri = Uri.parse(url)
            val scheme = uri.scheme ?: return null
            val host = uri.host ?: return scheme
            val port = uri.port
            if (port >= 0) "$scheme://$host:$port" else "$scheme://$host"
        } catch (_: Exception) {
            null
        }
    }

    private fun resolveThemeBackgroundColor(): Int {
        val typedValue = TypedValue()
        val resolved = activity.theme.resolveAttribute(android.R.attr.colorBackground, typedValue, true)
        if (!resolved) {
            return Color.WHITE
        }
        return if (typedValue.resourceId != 0) {
            ContextCompat.getColor(activity, typedValue.resourceId)
        } else {
            typedValue.data
        }
    }
}

File: apps/iris/src-tauri/plugins/mobile-browser/ios/Sources/MobileBrowserPlugin.swift
import Foundation
import Tauri
import UIKit
import WebKit

struct CreateArgs: Decodable {
  let label: String
  let url: String
  let x: Double
  let y: Double
  let width: Double
  let height: Double
  let scale: Double
  let initScript: String
  let diagnosticScript: String
  let allowedOriginRule: String?
  let actualUrlRoot: String?
}

struct LabelArgs: Decodable {
  let label: String
}

struct NavigateArgs: Decodable {
  let label: String
  let url: String
}

struct BoundsArgs: Decodable {
  let label: String
  let x: Double
  let y: Double
  let width: Double
  let height: Double
  let scale: Double
}

struct HistoryArgs: Decodable {
  let label: String
  let direction: String
}

struct ShellOverlayArgs: Decodable {
  let enabled: Bool
  let x: Double
  let y: Double
  let width: Double
  let height: Double
  let scale: Double
}

private final class BrowserEntry: NSObject, WKNavigationDelegate {
  let args: CreateArgs
  let webView: WKWebView
  let plugin: MobileBrowserPlugin

  init(args: CreateArgs, plugin: MobileBrowserPlugin) {
    self.args = args
    self.plugin = plugin

    let configuration = WKWebViewConfiguration()
    configuration.websiteDataStore = .default()
    configuration.userContentController = WKUserContentController()
    configuration.userContentController.addUserScript(
      WKUserScript(
        source: args.initScript,
        injectionTime: .atDocumentStart,
        forMainFrameOnly: true
      )
    )

    self.webView = WKWebView(frame: .zero, configuration: configuration)
    super.init()
    self.webView.navigationDelegate = self
    self.webView.backgroundColor = UIColor(red: 15 / 255, green: 15 / 255, blue: 15 / 255, alpha: 1)
    self.webView.isOpaque = true
  }

  func webView(_ webView: WKWebView, decidePolicyFor navigationAction: WKNavigationAction, decisionHandler: @escaping (WKNavigationActionPolicy) -> Void) {
    guard navigationAction.targetFrame?.isMainFrame != false else {
      decisionHandler(.allow)
      return
    }

    let url = navigationAction.request.url?.absoluteString ?? ""
    if plugin.shouldInterceptNavigation(args: args, url: url) {
      plugin.emitLocation(label: args.label, url: url, source: "navigation")
      decisionHandler(.cancel)
      return
    }

    decisionHandler(.allow)
  }

  func webView(_ webView: WKWebView, didStartProvisionalNavigation navigation: WKNavigation!) {
    guard let url = webView.url?.absoluteString else { return }
    plugin.emitLocation(label: args.label, url: url, source: "navigation")
    plugin.emitPageLoad(label: args.label, url: url, event: "started")
  }

  func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
    guard let url = webView.url?.absoluteString else { return }
    if plugin.shouldInjectScripts(args: args, url: url) {
      plugin.injectScripts(on: webView, args: args)
    }
    plugin.emitPageLoad(label: args.label, url: url, event: "finished")
  }

  func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
    plugin.emitDiagnostic(label: args.label, url: webView.url?.absoluteString, source: "navigation-error", error: error.localizedDescription)
  }

  func webView(_ webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!, withError error: Error) {
    plugin.emitDiagnostic(label: args.label, url: webView.url?.absoluteString ?? args.url, source: "provisional-navigation-error", error: error.localizedDescription)
  }
}

class MobileBrowserPlugin: Plugin {
  private var browsers = [String: BrowserEntry]()
  private weak var shellWebView: WKWebView?

  override func load(webview: WKWebView) {
    DispatchQueue.main.async {
      self.shellWebView = webview
    }
  }

  @objc public func create(_ invoke: Invoke) throws {
    let args = try invoke.parseArgs(CreateArgs.self)

    DispatchQueue.main.async {
      self.destroyBrowser(label: args.label)

      let entry = BrowserEntry(args: args, plugin: self)
      entry.webView.frame = self.frame(from: args)
      if let hostView = self.contentHostView() {
        hostView.addSubview(entry.webView)
        hostView.bringSubviewToFront(entry.webView)
      }
      self.browsers[args.label] = entry
      if let url = URL(string: args.url) {
        entry.webView.load(URLRequest(url: url))
      }
      invoke.resolve()
    }
  }

  @objc public func close(_ invoke: Invoke) throws {
    let args = try invoke.parseArgs(LabelArgs.self)
    DispatchQueue.main.async {
      self.destroyBrowser(label: args.label)
      invoke.resolve()
    }
  }

  @objc public func navigate(_ invoke: Invoke) throws {
    let args = try invoke.parseArgs(NavigateArgs.self)
    DispatchQueue.main.async {
      guard let entry = self.browsers[args.label], let url = URL(string: args.url) else {
        invoke.reject("Webview \(args.label) not found")
        return
      }
      entry.webView.load(URLRequest(url: url))
      invoke.resolve()
    }
  }

  @objc public func setBounds(_ invoke: Invoke) throws {
    let args = try invoke.parseArgs(BoundsArgs.self)
    DispatchQueue.main.async {
      guard let entry = self.browsers[args.label] else {
        invoke.reject("Webview \(args.label) not found")
        return
      }
      entry.webView.frame = self.frame(from: args)
      self.contentHostView()?.bringSubviewToFront(entry.webView)
      invoke.resolve()
    }
  }

  @objc public func setShellOverlay(_ invoke: Invoke) throws {
    _ = try invoke.parseArgs(ShellOverlayArgs.self)
    DispatchQueue.main.async {
      invoke.resolve()
    }
  }

  @objc public func history(_ invoke: Invoke) throws {
    let args = try invoke.parseArgs(HistoryArgs.self)
    DispatchQueue.main.async {
      guard let entry = self.browsers[args.label] else {
        invoke.reject("Webview \(args.label) not found")
        return
      }
      switch args.direction {
      case "back":
        if entry.webView.canGoBack { entry.webView.goBack() }
      case "forward":
        if entry.webView.canGoForward { entry.webView.goForward() }
      default:
        invoke.reject("Invalid history direction")
        return
      }
      invoke.resolve()
    }
  }

  @objc public func reload(_ invoke: Invoke) throws {
    let args = try invoke.parseArgs(LabelArgs.self)
    DispatchQueue.main.async {
      guard let entry = self.browsers[args.label] else {
        invoke.reject("Webview \(args.label) not found")
        return
      }
      entry.webView.reload()
      invoke.resolve()
    }
  }

  @objc public func currentUrl(_ invoke: Invoke) throws {
    let args = try invoke.parseArgs(LabelArgs.self)
    DispatchQueue.main.async {
      guard let entry = self.browsers[args.label] else {
        invoke.reject("Webview \(args.label) not found")
        return
      }
      invoke.resolve(["url": entry.webView.url?.absoluteString])
    }
  }

  func emitLocation(label: String, url: String, source: String) {
    trigger("location", data: [
      "label": label,
      "url": url,
      "source": source,
    ])
  }

  func emitPageLoad(label: String, url: String, event: String) {
    trigger("page-load", data: [
      "label": label,
      "url": url,
      "event": event,
    ])
  }

  func emitDiagnostic(label: String, url: String?, source: String, error: String) {
    trigger("diagnostic", data: [
      "label": label,
      "url": url ?? NSNull(),
      "source": source,
      "error": error,
    ])
  }

  func shouldInterceptNavigation(args: CreateArgs, url: String) -> Bool {
    if url.hasPrefix("htree://") {
      return true
    }

    if let actualUrlRoot = args.actualUrlRoot {
      return !url.hasPrefix(actualUrlRoot)
    }

    guard let allowedOriginRule = args.allowedOriginRule else {
      return false
    }

    return origin(of: url) != allowedOriginRule
  }

  func shouldInjectScripts(args: CreateArgs, url: String) -> Bool {
    if let actualUrlRoot = args.actualUrlRoot {
      return url.hasPrefix(actualUrlRoot)
    }

    guard let allowedOriginRule = args.allowedOriginRule else {
      return true
    }

    return origin(of: url) == allowedOriginRule
  }

  func injectScripts(on webView: WKWebView, args: CreateArgs) {
    webView.evaluateJavaScript(args.initScript, completionHandler: nil)
    webView.evaluateJavaScript(args.diagnosticScript, completionHandler: nil)
    DispatchQueue.main.asyncAfter(deadline: .now() + 0.15) {
      webView.evaluateJavaScript(args.initScript, completionHandler: nil)
      webView.evaluateJavaScript(args.diagnosticScript, completionHandler: nil)
    }
    DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
      webView.evaluateJavaScript(args.initScript, completionHandler: nil)
      webView.evaluateJavaScript(args.diagnosticScript, completionHandler: nil)
    }
  }

  private func destroyBrowser(label: String) {
    guard let entry = browsers.removeValue(forKey: label) else {
      return
    }
    entry.webView.navigationDelegate = nil
    entry.webView.stopLoading()
    entry.webView.removeFromSuperview()
  }

  private func rootView() -> UIView? {
    return manager.viewController?.view
  }

  private func contentHostView() -> UIView? {
    return shellWebView?.superview ?? rootView()
  }

  private func frame(from args: CreateArgs) -> CGRect {
    CGRect(x: args.x, y: args.y, width: max(args.width, 0), height: max(args.height, 0))
  }

  private func frame(from args: BoundsArgs) -> CGRect {
    CGRect(x: args.x, y: args.y, width: max(args.width, 0), height: max(args.height, 0))
  }

  private func origin(of urlString: String) -> String? {
    guard let components = URLComponents(string: urlString), let scheme = components.scheme else {
      return nil
    }
    guard let host = components.host else {
      return scheme
    }
    if let port = components.port {
      return "\(scheme)://\(host):\(port)"
    }
    return "\(scheme)://\(host)"
  }
}

@_cdecl("init_plugin_mobile_browser")
func initPlugin() -> Plugin {
  return MobileBrowserPlugin()
}
